cupy/random/_generator.py

```diff
@@ -672,50 +672,72 @@ def _interval(self, mx, size):
-            curand.generate(
-                self._generator, sample.data.ptr, size32)
-            # Drop the samples that exceed the upper limit
+        sample = self._curand_generate(n_sample, dtype)
+
+        mx1 = mx + 1
+        if mx1 != (1 << (mx1.bit_length() - 1)):
+            # Get index of samples that exceed the upper limit
+            ng_indices = self._get_indices(sample, upper_limit, False)
+            n_ng = ng_indices.size
+
+            while n_ng > 0:
+                n_supplement = max(n_ng * 2, 1024)
+                supplement = self._curand_generate(n_supplement, dtype)
+
+                # Get index of supplements that are within the upper limit
+                ok_indices = self._get_indices(supplement, upper_limit, True)
+                n_ok = ok_indices.size
+
+                # Replace the values that exceed the upper limit
+                if n_ok >= n_ng:
+                    sample[ng_indices] = supplement[ok_indices[:n_ng]]
+                    n_ng = 0
+                else:
+                    sample[ng_indices[:n_ok]] = supplement[ok_indices]
+                    ng_indices = ng_indices[n_ok:]
+                    n_ng -= n_ok
+            sample %= mx1
+        else:
+            mask = (1 << mx.bit_length()) - 1
             sample &= mask
-            success = sample <= mx
 
-            if ret is None:
-                # If the sampling has finished in the first iteration,
-                # just return the sample.
-                if success.all():
-                    n_rem = 0
-                    ret = sample
-                    break
+        return sample.reshape(size)
 
-                # Allocate the return array.
-                ret = cupy.empty((n,), dtype=dtype)
-
-            n_succ = min(n_rem, int(success.sum()))
-            ret[n - n_rem:n - n_rem + n_succ] = sample[success][:n_succ]
-            n_rem -= n_succ
-
-        assert n_rem == 0
+    def _curand_generate(self, num, dtype):
+        sample = cupy.empty((num,), dtype=dtype)
+        # Call 32-bit RNG to fill 32-bit or 64-bit `sample`
+        size32 = sample.view(dtype=numpy.uint32).size
+        curand.generate(self._generator, sample.data.ptr, size32)
+        return sample
 
-        return ret.reshape(size)
+    def _get_indices(self, sample, upper_limit, cond):
+        dtype = numpy.uint32 if sample.size < 2**32 else numpy.uint64
+        flags = (sample <= upper_limit) if cond else (sample > upper_limit)
+        csum = cupy.cumsum(flags, dtype=dtype)
+        del flags
+        indices = cupy.empty((int(csum[-1]),), dtype=dtype)
+        self._kernel_get_indices(csum, indices, size=csum.size)
+        return indices
+
+    _kernel_get_indices = core.ElementwiseKernel(
+        'raw U csum', 'raw U indices',
+        '''
+        int j = 0;
+        if (i > 0) { j = csum[i-1]; }
+        if (csum[i] > j) { indices[j] = i; }
+        ''',
+        'cupy_get_indices')
```

Comment:
On `_get_indices` and its kernel, please consider contributing them to `cupy/core/_routines_indexing.pyx` (`cupy.nonzero`, `cupy.argwhere`).

Reply:
I got it.